[RHELC-884] Disable RHEL repos when performing checks

hosekadam opened the pull request:
<!-- Write a description of what the PR solves and how -->
We need to disable RHEL repos. If user registers the system to satellite which provides Centos and RHEL repos and the RHEL repos would be enabled, it would cause a fail. The recommended way is to disable the rhel repos prior the conversion, but is expected it might be forgotten. This helps avoiding the problems caused by wrong setup of the repos.

<!-- Link to relevant Red Hat Jira issues -->
Jira Issues: https://issues.redhat.com/browse/RHELC-884

<!-- List below in format of [RHELC-](https://issues.redhat.com/browse/RHELC-) -->
-

Checklist

- [ ] PR has been tested manually in a VM (either author or reviewer)
- [ ] Jira issue has been made public if possible
- [ ] `[RHELC-]` is part of the PR title <!-- For a proper sync with Jira -->
- [ ] GitHub label has been added to help with Release notes <!-- enhancement, bug-fix, no-changelog, security-hardening, breaking-change, test-coverage-enhancement -->
- [ ] PR title explains the change from the user's point of view
- [ ] Code and tests are documented properly
- [ ] The commits are squashed to as few commits as possible (without losing data) <!-- The commits can be squashed to 1 commit, but then we might lose data regarding moving something to a new file and then refactoring for example. Hence squash without losing data -->
- [ ] When merged: Jira issue has been updated to `Release Pending` if relevant

r0x0d commented:
@hosekadam, looking at how we are disabling the repos, maybe we should use `rhel*` instead of `rhel-*`? 

I'm asking this more because if RHEL introduces a new repo, or modifies the name at some point to have something like: `rhel9-something-foo-bar`, and this repository is enabled, then we will miss this repository by the time we are running our tools.

hosekadam commented:
@r0x0d Thanks for taking a look! I used `rhel-*` based on the names of the repos, But `rhel*` might be more generic and future-proof. I'll switch to that.

r0x0d commented on convert2rhel/pkghandler.py:
```diff
@@ -918,7 +918,16 @@ def _get_packages_to_update_yum():
+    # Disable the rhel repositories when checking if the packages are up to date.
+    # When user specifies rhel repos to use during conversion, disabling them is also needed.
+    # Note: Do not depend on the --no-rhsm option, enable repos can be specified without it.
+    disable_repos = ["rhel-*"] + tool_opts.enablerepo
```
```suggestion
    disable_repos = ["rhel*"] + tool_opts.enablerepo
```

r0x0d commented on convert2rhel/pkghandler.py:
```diff
@@ -935,6 +944,10 @@ def _get_packages_to_update_dnf(reposdir):
+    # Disable the rhel repositories when checking if the packages are up to date.
+    # When user specifies rhel repos to use during conversion, disabling them is also needed.
+    # Note: Do not depend on the --no-rhsm option, enable repos can be specified without it.
+    disable_repos = ["rhel-*"] + tool_opts.enablerepo
```
```suggestion
    disable_repos = ["rhel*"] + tool_opts.enablerepo
```

r0x0d commented on convert2rhel/actions/system_checks/is_loaded_kernel_latest.py:
```diff
@@ -48,6 +48,7 @@ def run(self):  # pylint: disable= too-many-return-statements
             "repoquery",
             "--setopt=exclude=",
             "--quiet",
+            "--disablerepo=rhel-*",
```
```suggestion
            "--disablerepo=rhel*",
```

hosekadam replied:
Thanks @r0x0d!

hosekadam commented on convert2rhel/unit_tests/pkghandler_test.py:
```diff
@@ -980,6 +980,105 @@ def test_get_packages_to_update_dnf(packages, reposdir, pretend_os, monkeypatch)
+    reason="No yum module detected on the system, skipping it.",
+)
+def test_get_packages_to_update_yum_rhel_repos(monkeypatch):
+    # Import the YUM inside the test. If YUM is not present the test isn't run
+    import yum
```
I'll take a look, there is one more problem with mocking the `doPackageList` - the tests failed. It was just a test if that `yum.repos.Repository` works :D

hosekadam replied:
Now it's working, the tests `test_get_packages_to_update_yum_rhel` and `test_get_packages_to_update_dnf_rhel` are ready @r0x0d

hosekadam commented:
So from the last time:
- unit tests for check if system is up-to-date were added for both yum and dnf
- changed the disable pattern from `rhel-*` to `rhel*`
- added the `--disablerepo=rhel*` into the unit tests for latest kernel
- @bocekm found there can be problem during backup, when the RHEL repos are enabled and the pkg to download has the same NEVRA in original vendor and rhel repos. There is possibility the RHEL one would be downloaded as backup of some pkg we need to backup (we are backuping just the special packages we are removing). I was looking also at the `RestorablePackageSet` but I don't think we wan't to disable RHEL repos there, I think they are needed in this place.

I'm not sure if there is requirement to add unit tests for `RestorablePackage.enable()` (right now we are just checking if `utils.download_pkg` was called) and for `IsLoadedKernelLatest.run()` (there I added the argument to the list of expected args).

SpyTec commented on convert2rhel/unit_tests/pkghandler_test.py:
```diff
@@ -980,6 +980,104 @@ def test_get_packages_to_update_dnf(packages, reposdir, pretend_os, monkeypatch)
+    assert rhel_repo_id in enabled_repos
+
+    # Choose one of the enabled repos on the system
+    # Avoid getting the rhel_repo_id if on 0 index, then get the repo id from 1
```
Why?

hosekadam replied:
I'm trying to disable one of the repos available on the system to disabling repo provided through the tool_opts (`monkeypatch.setattr(tool_opts, "enablerepo", repo_for_disable)`). I want to have something different than the added RHEL repo but still present on the system. 

So I'm getting it in this way - I know if on the 0 index is rhel repo, on the 1 will be something different I can use

SpyTec replied:
I'd write that out a bit clearer in the text, avoid getting something is clear but not super clear in the comments of why you do it

hosekadam replied:
Comment about it added.

SpyTec commented on convert2rhel/unit_tests/pkghandler_test.py:
```diff
@@ -980,6 +980,104 @@ def test_get_packages_to_update_dnf(packages, reposdir, pretend_os, monkeypatch)
+    # When user adds a repo through the --enable-repo option, it should be disabled
+    monkeypatch.setattr(tool_opts, "enablerepo", repo_for_disable)
+
+    # Patch the original YUM Base to ours to have access to it and contain the added repo
+    monkeypatch.setattr(pkgmanager, "YumBase", mock.Mock(return_value=base))
```
Good approach, that means we still test with the Yum/DnfBase still and not just plainly mocking it out

hosekadam replied:
Yes, exactly

SpyTec reviewed (commented):
LGTM, but confused about the skip if index 0 part

r0x0d commented on convert2rhel/pkghandler.py:
```diff
@@ -935,6 +944,10 @@ def _get_packages_to_update_dnf(reposdir):
+    # Disable the rhel repositories when checking if the packages are up to date.
+    # When user specifies rhel repos to use during conversion, disabling them is also needed.
+    # Note: Do not depend on the --no-rhsm option, enable repos can be specified without it.
+    disable_repos = ["rhel*"] + tool_opts.enablerepo
```
One tip, that may not be 100% necessary here (and definitely not a blocker), but you could move the disable_repos to be a parameter in the function, this way you only have to identify the rhel repos + tool_opts.enablerepo once in the caller, and pass this down to the functions.

hosekadam replied:
Yeah, it would help with decreasing duplicity of code. I'll do it, I like the idea. And as I'm looking on it, I see I forgot to disable the `tool_opts.enablerepo` on other places.

r0x0d reviewed (approved):
Looks good!

hosekadam commented:
- rebased

hosekadam commented:
@pr-watson needs to try something. The PR is now WIP

r0x0d commented on convert2rhel/backup/packages.py:
```diff
@@ -143,11 +144,15 @@ def enable(self):
                 loggerinst.info("The repository directory %s seems to be empty or non-existent.")
                 self.reposdir = None
 
+        disable_repos = ["rhel*"] + tool_opts.enablerepo
```
I think we should move this to the __init__ to allow the caller to override the disable_repos if needed. This way, we will always assume that rhel*+ enablerepo will be disabled for all package backup, regardless of what we choose. 

Maybe we could improve it like this: 

```python
def __init__(self, pkgs, reposdir=None, set_releasever=False, custom_releasever=None, varsdir=None, disable_repos=None):
        """
        Keep control of system packages before their removal to backup and
        restore in case of rollback.
        :param pkgs list[str]: List of packages to backup.
        :param reposdir str: If a custom repository directory location needs to
            be used, this parameter can be set with the location for it.
        :param set_releasever bool: If there is need to set the relesever while
            downloading the package with yumdownloader.
        :param custom_releasever str: Custom releasever in case it need to be
            overwritten and it differs from the `py:system_info.releasever`.
        :param varsdir str: Location to the variables directory in case the
            repository files needs to interpolate variables from those folders.
        :param disable_repos list[str]: Repositories to disable during package download. Defaults to None.
        """
        super(RestorablePackage, self).__init__()
        self.pkgs = pkgs
        self.reposdir = reposdir
        self.set_releasever = set_releasever
        self.custom_releasever = custom_releasever
        self.varsdir = varsdir
        self.disable_repos = disable_repos or ["rhel*"] + tool_opts.enablerepo
        self._backedup_pkgs_paths = []
```

Then later while we are downloading the packages, it would use it like this: 

```python
        for pkg in self.pkgs:
            self._backedup_pkgs_paths.append(
                utils.download_pkg(
                    pkg=pkg,
                    dest=BACKUP_DIR,
                    # RHELC-884 disable the RHEL repos to avoid downloading pkg from them
                    disable_repos=self.disable_repos,
                    set_releasever=self.set_releasever,
                    custom_releasever=self.custom_releasever,
                    varsdir=self.varsdir,
                    reposdir=self.reposdir,
                )
            )
```

r0x0d commented on convert2rhel/pkghandler.py:
```diff
@@ -900,25 +900,35 @@ def get_total_packages_to_update(reposdir):
 
 
-def _get_packages_to_update_yum():
+def _get_packages_to_update_yum(disable_repos=[]):
     """Use yum to get all the installed packages that have an update available.
 
     :return: Return a list of packages that needs to be updated.
     :rtype: list[str] | list
     """
     all_packages = []
```
```suggestion
def _get_packages_to_update_yum(disable_repos=None):
    """Use yum to get all the installed packages that have an update available.

    :return: Return a list of packages that needs to be updated.
    :rtype: list[str] | list
    """
     disable_repos = disable_repos or []
    all_packages = []
```

hosekadam commented:
- created `repo.get_rhel_repos_to_disable()` to reduce duplicity
- moved preparation of the disablerepo list higher in the chain in many places to reduce duplicity

TBD:
- one bug waiting to be solved in `_get_package_repositories`, the repoquery should contain the `"--disablerepo=rhel*"` (pushing the rest for having backup), tested with it, needs unit tests and better implementation
- solve the conflicts :(

hosekadam commented:
- rebased

hosekadam commented:
-  _get_package_repositories, the repoquery now contains the "--disablerepo=rhel*"

hosekadam commented on convert2rhel/pkghandler.py:
```diff
@@ -393,8 +393,11 @@ def _get_package_repositories(pkgs):
         query_format = "C2R %{NAME}-%{EPOCH}:%{VERSION}-%{RELEASE}.%{ARCH}&%{REPOID}\n"
 
+    # RHELC-884 disable the RHEL repos to avoid downloading pkgs from them.
+    disable_repo_command = repo.get_rhel_disable_repos_command()
```
This is part of discussion. WIP

hosekadam commented:
- refactor how the disable_repos is passed
- disabled RHEL repos when retrieving repo info from pkgs in `RemoveSpecialPkgs` and `ListThirdPartyPackages`
